### ISSUE TITLE
Improve autowiring isFresh() performance by touch()ing the compiled container.

### ISSUE THREAD

**Issue by d-ph** [symfony/symfony#35724]
**Description**

Hello,

An autowired class is checked for freshness by [ReflectionClassResource.php](https://github.com/symfony/symfony/blob/f46ab58bcff6305331678892c3b4f9f3cf34a15d/src/Symfony/Component/Config/Resource/ReflectionClassResource.php). The check is twofold:

1. Compare the "observed" file's `filemtime()` with the compiled container file's `filemtime()` [(source code)](https://github.com/symfony/symfony/blob/f46ab58bcff6305331678892c3b4f9f3cf34a15d/src/Symfony/Component/Config/Resource/ReflectionClassResource.php#L53). If the observed file has been modified after the container was compiled, perform the second check.
2. Reflect the observed file to retrieve the php class in it and perform a more exhaustive check.

The first check is quick and the second is fast enough. However, it's best if the first check is always conclusive, especially when there are dozens of classes being autowired.

The issue is that a mere edit to a method's body in an autowired class makes the first check not conclusive anymore, and from that point on, the class will be thoroughly reflected on every http request until something finally kicks off container compilation (and this might take a while).

What would you say if Symfony was `touch()`ing the complied container file every time the overall `isFresh()` check concludes that the container is fresh? By making the following change, the issue described above is resolved:
```php
// https://github.com/symfony/symfony/blob/f46ab58bcff6305331678892c3b4f9f3cf34a15d/src/Symfony/Component/Config/ResourceCheckerConfigCache.php#L93

$time = filemtime($this->file);

foreach ($meta as $resource) {
    /* @var ResourceInterface $resource */
    foreach ($this->resourceCheckers as $checker) {
        if (!$checker->supports($resource)) {
            continue; // next checker
        }
        if ($checker->isFresh($resource, $time)) {
            break; // no need to further check this resource
        }

        return false; // cache is stale
    }
    // no suitable checker found, ignore this resource
}

// ------------------ THE FOLLOWING LINE IS NEW
touch($this->file);

return true;
```

Note that this `touch()` would not affect the performance of webapps that run in non-debug mode (i.e. when they run in the production environment). It could, however, affect performance when the debug mode is on and when there are many concurrent requests being served. Each would attempt to `touch()` the container file which, in high traffic scenarios, would compete for I/O to this file (note that this would be quite negligible).

The issue with the suggested solution is of "best practice" nature. One might not like side effects to happen in an is*() method.

So again: may I have your opinion about the suggested solution, please? I can PR if it's all good.

Thank you.

**Comment by nicolas-grekas** [comment]
I'd be interested in a PR and a few numbers to support the claim!

**Comment by d-ph** [comment]
Hi,

I tested the performance with and without the suggested `touch()` solution on the `symfony/symfony-demo` homepage and I concluded that the speed-up is not clearly visible. I edited about 10 of the autowired classes to have a dummy change in them and then kept reloading the homepage to get the hang of the numbers, and then compared the numbers. It might be a placebo but I felt like I get under 100ms responses more consistently with the suggested solution. But overall, I wouldn't say the solution is clearly and visibly improving the performance (e.g. more than 100ms difference).

The time spent on reflecting the file is depending on various things, like the number of methods in a class (Doctrine entity classes usually have a lot of setters and getters) or the amount of docblocks. I don't have a real-life project that uses Symfony autowiring, so I can't really tell whether the suggested solution is any different under less contrived testing.

At this point, I'm ok if the ticket is closed if there's no further desire to action it. I might come back to it when I have better numbers to back up my claims.

Thank you.
